www/src/components/repository/RepositoryEdit.tsx

```diff
@@ -0,0 +1,677 @@
+import React, {
+  EventHandler,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react'
+import { ApolloError, useMutation, useQuery } from '@apollo/client'
+import {
+  Div,
+  Flex,
+  Form,
+  H2,
+  Img,
+  MenuItem,
+  P,
+  Select,
+  Switch,
+} from 'honorable'
+import { TextInput } from 'grommet'
+import {
+  Button,
+  CloseIcon,
+  FormField,
+  Input,
+  SearchIcon,
+  Tag,
+} from 'pluralsh-design-system'
+import isEqual from 'lodash/isEqual'
+import isArray from 'lodash/isArray'
+import uniqWith from 'lodash/uniqWith'
+import { useFilePicker } from 'react-sage'
+import { filter } from 'lodash'
+import styled from '@emotion/styled'
+
+import RepositoryContext from '../../contexts/RepositoryContext'
+import { capitalize } from '../../utils/string'
+import { generatePreview } from '../../utils/file'
+import { AuthMethod as authMethods } from '../oidc/types'
+
+import {
+  REPOSITORY_QUERY,
+  TAGS_SEARCH_QUERY,
+  UPDATE_REPOSITORY_MUTATION,
+} from './queries'
+
+export const categories = [
+  'DEVOPS',
+  'DATABASE',
+  'MESSAGING',
+  'SECURITY',
+  'DATA',
+  'PRODUCTIVITY',
+  'NETWORK',
+  'STORAGE',
+]
+
+const StyledTextInput = styled(TextInput)`
+  border: 1px solid #454954;
+  border-radius: 3px;
+  &:focus {
+    border-color: #5c77ff;
+  }
+`
+
+function useUpdateState<T extends { [key: string]: unknown }>(initialState: T) {
+  const [state, setState] = useState({ ...initialState })
+
+  const update = useCallback(
+    (update: Partial<T>) => {
+      setState({ ...state, ...update })
+    },
+    [state]
+  )
+  const reset = useCallback(() => {
+    setState({ ...initialState })
+  }, [initialState])
+
+  const hasUpdates = useMemo(() => {
+    for (const [prop, value] of Object.entries(state)) {
+      const curVal = !value && typeof value === 'string' ? null : value
+      const initialVal =
+        !initialState[prop] && typeof value === 'string'
+          ? null
+          : initialState[prop]
+      if (!isEqual(curVal, initialVal)) {
```

Comment:
Wouldn't using lodash `isEqual` in general work here? It can do a deep object comparison.

Reply:
I had a bunch of nulls in initialState that I needed to match empty strings, but I fixed by just making sure initialState started with empty strings.